### ISSUE TITLE
关于使用自己的数据集

### ISSUE THREAD

**Issue by chostorjax** [thuml/Autoformer#36]
作者你好！非常感谢您的这篇论文给我带来了启发与灵感，在使用我自己的数据集的时候我遇到了一些问题，希望您能百忙之中抽空为我解答一下。

这是我的数据集：
[AL_WIND_07_12.csv](https://github.com/thuml/Autoformer/files/8536651/AL_WIND_07_12.csv)

这是运行结果：Args in experiment:
Namespace(activation='gelu', batch_size=32, bucket_size=4, c_out=6, checkpoints='./checkpoints/', d_ff=2048, d_layers=1, d_model=512, data='windresult(2)', data_path='windresult(2).csv', dec_in=6, des='exp', devices='0,1,2,3', distil=True, do_predict=False, dropout=0.05, e_layers=2, embed='timeF', enc_in=6, factor=3, features='M', freq='t', gpu=0, is_training=1, itr=1, label_len=48, learning_rate=1e-05, loss='mse', lradj='type1', model='Autoformer', model_id='windresult(2)_96_168', moving_avg=25, n_hashes=4, n_heads=8, num_workers=10, output_attention=False, patience=3, pred_len=168, root_path='./data', seq_len=96, target='OT', train_epochs=10, use_amp=False, use_gpu=True, use_multi_gpu=False)
Use GPU: cuda:0
>>>>>>>start training : windresult(2)_96_168_Autoformer_windresult(2)_ftM_sl96_ll48_pl168_dm512_nh8_el2_dl1_df2048_fc3_ebtimeF_dtTrue_exp_0>>>>>>>>>>>>>>>>>>>>>>>>>>
train 48860
val 6851
test 13868
	iters: 100, epoch: 1 | loss: 1.3873028
	speed: 0.2157s/iter; left time: 3270.9135s
	iters: 200, epoch: 1 | loss: 5.2782578
	speed: 0.1973s/iter; left time: 2971.3670s
	iters: 300, epoch: 1 | loss: 0.8285253
	speed: 0.1991s/iter; left time: 2978.6951s
	iters: 400, epoch: 1 | loss: 0.6783259
	speed: 0.2014s/iter; left time: 2993.2308s
	iters: 500, epoch: 1 | loss: 0.5328820
	speed: 0.2033s/iter; left time: 3001.2006s
	iters: 600, epoch: 1 | loss: 0.4745004
	speed: 0.2032s/iter; left time: 2978.4878s
	iters: 700, epoch: 1 | loss: 0.6617956
	speed: 0.2042s/iter; left time: 2973.3634s
	iters: 800, epoch: 1 | loss: 0.8940198
	speed: 0.2053s/iter; left time: 2969.4257s
	iters: 900, epoch: 1 | loss: 0.4608745
	speed: 0.2061s/iter; left time: 2959.5262s
	iters: 1000, epoch: 1 | loss: 0.6747336
	speed: 0.2064s/iter; left time: 2942.9809s
	iters: 1100, epoch: 1 | loss: 1.0136081
	speed: 0.2069s/iter; left time: 2929.6766s
	iters: 1200, epoch: 1 | loss: 0.5815385
	speed: 0.2067s/iter; left time: 2906.8596s
	iters: 1300, epoch: 1 | loss: 0.7630769
	speed: 0.2069s/iter; left time: 2889.0642s
	iters: 1400, epoch: 1 | loss: 0.9147382
	speed: 0.2073s/iter; left time: 2872.9977s
	iters: 1500, epoch: 1 | loss: 0.5026098
	speed: 0.2068s/iter; left time: 2845.5929s
Epoch: 1 cost time: 313.20935106277466
Epoch: 1, Steps: 1526 | Train Loss: 1.2920112 Vali Loss: 0.6382990 Test Loss: 1.6633676
Validation loss decreased (inf --> 0.638299).  Saving model ...
Updating learning rate to 1e-05
	iters: 100, epoch: 2 | loss: 3.1819613
	speed: 1.0378s/iter; left time: 14150.4108s
	iters: 200, epoch: 2 | loss: 0.6240104
	speed: 0.2074s/iter; left time: 2806.6792s
	iters: 300, epoch: 2 | loss: 0.4239972
	speed: 0.2075s/iter; left time: 2787.2130s
	iters: 400, epoch: 2 | loss: 0.5790454
	speed: 0.2073s/iter; left time: 2764.3452s
	iters: 500, epoch: 2 | loss: 0.5129109
	speed: 0.2072s/iter; left time: 2742.8357s
	iters: 600, epoch: 2 | loss: 3.0552266
	speed: 0.2072s/iter; left time: 2722.1359s
	iters: 700, epoch: 2 | loss: 0.6153819
	speed: 0.2074s/iter; left time: 2702.9947s
	iters: 800, epoch: 2 | loss: 0.6108029
	speed: 0.2075s/iter; left time: 2683.7782s
	iters: 900, epoch: 2 | loss: 1.0149522
	speed: 0.2075s/iter; left time: 2663.1304s
	iters: 1000, epoch: 2 | loss: 0.7320828
	speed: 0.2075s/iter; left time: 2642.3782s
	iters: 1100, epoch: 2 | loss: 0.6959580
	speed: 0.2074s/iter; left time: 2620.9557s
	iters: 1200, epoch: 2 | loss: 0.4868296
	speed: 0.2069s/iter; left time: 2593.7263s
	iters: 1300, epoch: 2 | loss: 0.5166126
	speed: 0.2069s/iter; left time: 2572.9765s
	iters: 1400, epoch: 2 | loss: 2.1325247
	speed: 0.2074s/iter; left time: 2557.7465s
	iters: 1500, epoch: 2 | loss: 0.5191646
	speed: 0.2069s/iter; left time: 2531.3638s
Epoch: 2 cost time: 317.8026850223541
Epoch: 2, Steps: 1526 | Train Loss: 1.2491093 Vali Loss: 0.6379986 Test Loss: 1.6479515
Validation loss decreased (0.638299 --> 0.637999).  Saving model ...
Updating learning rate to 5e-06
	iters: 100, epoch: 3 | loss: 4.1040535
	speed: 1.0396s/iter; left time: 12587.9771s
	iters: 200, epoch: 3 | loss: 4.2198677
	speed: 0.2125s/iter; left time: 2551.6086s
	iters: 300, epoch: 3 | loss: 0.5374264
	speed: 0.2089s/iter; left time: 2488.3029s
	iters: 400, epoch: 3 | loss: 0.6558690
	speed: 0.2074s/iter; left time: 2449.3538s
	iters: 500, epoch: 3 | loss: 1.3005149
	speed: 0.2074s/iter; left time: 2428.3376s
	iters: 600, epoch: 3 | loss: 0.6544358
	speed: 0.2079s/iter; left time: 2413.1934s
	iters: 700, epoch: 3 | loss: 5.3746767
	speed: 0.2080s/iter; left time: 2394.1207s
	iters: 800, epoch: 3 | loss: 0.5379453
	speed: 0.2110s/iter; left time: 2407.4353s
	iters: 900, epoch: 3 | loss: 0.5085211
	speed: 0.2121s/iter; left time: 2399.0955s
	iters: 1000, epoch: 3 | loss: 2.3799009
	speed: 0.2124s/iter; left time: 2381.1714s
	iters: 1100, epoch: 3 | loss: 0.6582829
	speed: 0.2122s/iter; left time: 2357.6999s
	iters: 1200, epoch: 3 | loss: 0.6864679
	speed: 0.2121s/iter; left time: 2334.9807s
	iters: 1300, epoch: 3 | loss: 0.5495363
	speed: 0.2124s/iter; left time: 2317.5838s
	iters: 1400, epoch: 3 | loss: 0.4320998
	speed: 0.2126s/iter; left time: 2297.7912s
	iters: 1500, epoch: 3 | loss: 1.1172121
	speed: 0.2122s/iter; left time: 2271.9986s
Epoch: 3 cost time: 322.95352268218994
Epoch: 3, Steps: 1526 | Train Loss: 1.2290818 Vali Loss: 0.6389777 Test Loss: 1.6318545
EarlyStopping counter: 1 out of 3
Updating learning rate to 2.5e-06
	iters: 100, epoch: 4 | loss: 6.5205026
	speed: 1.1262s/iter; left time: 11918.5769s
	iters: 200, epoch: 4 | loss: 0.5527835
	speed: 0.2125s/iter; left time: 2227.8205s
	iters: 300, epoch: 4 | loss: 0.7205955
	speed: 0.2125s/iter; left time: 2206.8931s
	iters: 400, epoch: 4 | loss: 0.5857182
	speed: 0.2126s/iter; left time: 2185.6987s
	iters: 500, epoch: 4 | loss: 1.8298072
	speed: 0.2127s/iter; left time: 2165.4811s
	iters: 600, epoch: 4 | loss: 3.0394719
	speed: 0.2128s/iter; left time: 2145.9223s
	iters: 700, epoch: 4 | loss: 2.4748251
	speed: 0.2134s/iter; left time: 2129.9469s
	iters: 800, epoch: 4 | loss: 0.7085900
	speed: 0.2142s/iter; left time: 2117.3976s
	iters: 900, epoch: 4 | loss: 0.6424207
	speed: 0.2127s/iter; left time: 2081.0393s
	iters: 1000, epoch: 4 | loss: 0.7736093
	speed: 0.2137s/iter; left time: 2068.8437s
	iters: 1100, epoch: 4 | loss: 0.5847422
	speed: 0.2156s/iter; left time: 2065.7773s
	iters: 1200, epoch: 4 | loss: 0.5024280
	speed: 0.2156s/iter; left time: 2044.1739s
	iters: 1300, epoch: 4 | loss: 0.5836925
	speed: 0.2154s/iter; left time: 2021.3819s
	iters: 1400, epoch: 4 | loss: 0.7704700
	speed: 0.2155s/iter; left time: 2000.8996s
	iters: 1500, epoch: 4 | loss: 0.5966436
	speed: 0.2150s/iter; left time: 1974.5776s
Epoch: 4 cost time: 327.7894377708435
Epoch: 4, Steps: 1526 | Train Loss: 1.2139685 Vali Loss: 0.6468018 Test Loss: 1.6328001
EarlyStopping counter: 2 out of 3
Updating learning rate to 1.25e-06
	iters: 100, epoch: 5 | loss: 0.6045610
	speed: 1.1276s/iter; left time: 10212.5691s
	iters: 200, epoch: 5 | loss: 4.1248837
	speed: 0.2127s/iter; left time: 1905.1793s
	iters: 300, epoch: 5 | loss: 4.0812445
	speed: 0.2129s/iter; left time: 1885.6748s
	iters: 400, epoch: 5 | loss: 0.6445397
	speed: 0.2127s/iter; left time: 1862.5061s
	iters: 500, epoch: 5 | loss: 4.0005364
	speed: 0.2125s/iter; left time: 1839.6578s
	iters: 600, epoch: 5 | loss: 0.6750206
	speed: 0.2128s/iter; left time: 1820.5966s
	iters: 700, epoch: 5 | loss: 0.9796102
	speed: 0.2125s/iter; left time: 1797.1381s
	iters: 800, epoch: 5 | loss: 0.5318608
	speed: 0.2128s/iter; left time: 1778.5378s
	iters: 900, epoch: 5 | loss: 0.5962467
	speed: 0.2125s/iter; left time: 1754.2072s
	iters: 1000, epoch: 5 | loss: 3.0690393
	speed: 0.2125s/iter; left time: 1733.2038s
	iters: 1100, epoch: 5 | loss: 0.6859824
	speed: 0.2124s/iter; left time: 1711.4450s
	iters: 1200, epoch: 5 | loss: 0.6713046
	speed: 0.2121s/iter; left time: 1687.7346s
	iters: 1300, epoch: 5 | loss: 0.4987230
	speed: 0.2118s/iter; left time: 1663.7345s
	iters: 1400, epoch: 5 | loss: 0.6661511
	speed: 0.2116s/iter; left time: 1641.1604s
	iters: 1500, epoch: 5 | loss: 0.5974678
	speed: 0.2129s/iter; left time: 1630.4536s
Epoch: 5 cost time: 326.17496824264526
Epoch: 5, Steps: 1526 | Train Loss: 1.2040599 Vali Loss: 0.6492615 Test Loss: 1.6299148
EarlyStopping counter: 3 out of 3
Early stopping
>>>>>>>testing : windresult(2)_96_168_Autoformer_windresult(2)_ftM_sl96_ll48_pl168_dm512_nh8_el2_dl1_df2048_fc3_ebtimeF_dtTrue_exp_0<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
test 13868
test shape: (433, 32, 168, 6) (433, 32, 168, 6)
test shape: (13856, 168, 6) (13856, 168, 6)
mse:1.6479512453079224, mae:0.682569682598114

结果可视化之后：
![Figure_1](https://user-images.githubusercontent.com/102529524/164591897-6075168b-32e7-488a-af9f-5b54abe435c9.png)

个人猜想：可能是factors过大导致引入过多的噪音而产生这样的结果？

希望您能为我的问题提供一些解决思路，万分感谢!

**Comment by wuhaixu2016** [comment]
Hi, thanks for the interest.
I think you can further tune your parameters for better results following the issue https://github.com/thuml/Autoformer/issues/35

Here are some parameters you can take into consideration:
- kernel_size in moving_avg
- seq_len, label_len, pred_len
- factor

**Comment by chostorjax** [comment]
非常感谢您的建议，经过您的建议我的图得到了一定的提升！我想多问一下：按我个人的理解Autoformer的预测方式是给定一组数据，我要预测7天。（1）它是一次性把7天的数据都给预测出来呢？（2）还是类似于Lstm的单步预测一样，先预测第一天的值，紧接着再把第一天的值也归进数据里再进行预测第二天的值呢？

**Comment by Hymnhyz** [comment]
> 

显然是前者啊  非自回归模型会有误差累计  对于长期序列预测效果不好 并且也慢

**Comment by LY-must-win** [comment]
> moving_avg的大小如何设置最好

**Comment by deepConnectionism** [comment]
> > moving_avg的大小如何设置最好
> 
> 

实验来确定
